File: apps/android/app/src/main/kotlin/io/f7z/olas/core/OlasHaptics.kt
package io.f7z.olas.core

import android.content.Context
import android.media.AudioAttributes
import android.media.SoundPool
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.view.HapticFeedbackConstants
import android.view.View

// ---------------------------------------------------------------------------
// Haptics
// ---------------------------------------------------------------------------

/**
 * Thin wrappers around [View.performHapticFeedback] for Olas interaction semantics.
 *
 * The OS gates execution automatically — Android respects "Touch vibration" in
 * Settings → Sound & Vibration. No extra guarding is needed here.
 *
 * Usage in Compose: obtain [View] via `val view = LocalView.current` and pass it.
 */
object OlasHaptics {

    /** Light click — like/heart tap. */
    fun impactLight(view: View) {
        view.performHapticFeedback(HapticFeedbackConstants.KEYBOARD_TAP)
    }

    /** Rigid click — filter carousel per-cell selection tick. */
    fun impactRigid(view: View) {
        view.performHapticFeedback(HapticFeedbackConstants.KEYBOARD_TAP)
    }

    /** Soft thud — new-posts pill tap, revealing queued content. */
    fun impactSoft(view: View) {
        view.performHapticFeedback(HapticFeedbackConstants.VIRTUAL_KEY)
    }

    /** Selection tick — feed mode tab switch and new-posts pill appear. */
    fun selectionChanged(view: View) {
        view.performHapticFeedback(HapticFeedbackConstants.CLOCK_TICK)
    }

    /**
     * Success notification — zap sent or post published.
     * Uses [HapticFeedbackConstants.CONFIRM] on API 30+ for a triple-tap feel;
     * falls back to [HapticFeedbackConstants.VIRTUAL_KEY] on older releases.
     */
    fun notificationSuccess(view: View) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            view.performHapticFeedback(HapticFeedbackConstants.CONFIRM)
        } else {
            view.performHapticFeedback(HapticFeedbackConstants.VIRTUAL_KEY)
        }
    }
}

// ---------------------------------------------------------------------------
// Sound
// ---------------------------------------------------------------------------

/**
 * Opt-in sound effects for Olas, gated by a [SharedPreferences] boolean
 * `"sound_effects_enabled"` in the `"olas_prefs"` file (default `false`).
 *
 * Uses [AudioAttributes.USAGE_ASSISTANCE_SONIFICATION] so effects:
 * - respect the hardware ringer/vibration mode,
 * - do not interrupt background music.
 *
 * Audio assets (`shutter_soft.ogg`, `zap_chime.ogg`) live in `res/raw/` and
 * are **optional**. If an asset is absent the call is a silent no-op. Real files
 * can be added later without changing call-site code.
 */
object OlasSound {

    private const val PREFS_FILE = "olas_prefs"
    private const val KEY_ENABLED = "sound_effects_enabled"

    fun isEnabled(context: Context): Boolean =
        context.applicationContext
            .getSharedPreferences(PREFS_FILE, Context.MODE_PRIVATE)
            .getBoolean(KEY_ENABLED, false)

    fun setEnabled(context: Context, enabled: Boolean) {
        context.applicationContext
            .getSharedPreferences(PREFS_FILE, Context.MODE_PRIVATE)
            .edit()
            .putBoolean(KEY_ENABLED, enabled)
            .apply()
    }

    /** Soft camera shutter — plays on successful post publish. */
    fun shutterSoft(context: Context) {
        if (!isEnabled(context)) return
        playRaw(context, "shutter_soft")
    }

    /** Warm chime — plays when a zap is sent. */
    fun zapChime(context: Context) {
        if (!isEnabled(context)) return
        playRaw(context, "zap_chime")
    }

    // -------------------------------------------------------------------------

    /**
     * Loads and plays a `res/raw/<name>` asset via a short-lived [SoundPool].
     * The pool is released ~3 s after load completes (well beyond any UI sound duration).
     * Returns silently if the resource identifier is 0 (asset absent).
     */
    private fun playRaw(context: Context, name: String) {
        val appContext = context.applicationContext
        val resId = appContext.resources.getIdentifier(name, "raw", appContext.packageName)
        if (resId == 0) return // Asset not yet bundled — silent no-op by design.

        runCatching {
            val pool = SoundPool.Builder()
                .setMaxStreams(2)
                .setAudioAttributes(
                    AudioAttributes.Builder()
                        .setUsage(AudioAttributes.USAGE_ASSISTANCE_SONIFICATION)
                        .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                        .build()
                )
                .build()

            pool.setOnLoadCompleteListener { soundPool, sampleId, _ ->
                soundPool.play(sampleId, 1f, 1f, 0, 0, 1f)
                // Release after a generous delay to allow playback to finish.
                Handler(Looper.getMainLooper()).postDelayed({ soundPool.release() }, 3_000L)
            }
            pool.load(appContext, resId, 1)
        }
        // Silently swallow any SoundPool construction/load failure.
    }
}

File: apps/ios/Olas/Core/OlasHaptics.swift
import UIKit
import AVFoundation

// MARK: - Haptics

/// Centralised haptic helper for Olas.
///
/// All methods create a fresh feedback generator, call `prepare()` immediately
/// before triggering — keeping the Taptic Engine warm without long-lived state.
///
/// The OS gates execution automatically:
/// - `UIFeedbackGenerator` respects System Preferences → Sounds & Haptics → Vibration
///   and the Low Power Mode flag. No extra guarding is needed here.
/// - Calling these from the main actor (SwiftUI callbacks, UploadQueue) is always safe.
@MainActor
enum OlasHaptics {

    // MARK: Impact

    /// Light click — use for like/heart tap.
    static func impactLight() {
        let g = UIImpactFeedbackGenerator(style: .light)
        g.prepare()
        g.impactOccurred()
    }

    /// Soft thud — use for revealing queued content (new-posts pill tap).
    static func impactSoft() {
        let g = UIImpactFeedbackGenerator(style: .soft)
        g.prepare()
        g.impactOccurred()
    }

    /// Rigid click — use for filter carousel per-cell selection tick.
    static func impactRigid() {
        let g = UIImpactFeedbackGenerator(style: .rigid)
        g.prepare()
        g.impactOccurred()
    }

    // MARK: Notification

    /// Success triple-tap — use for zap sent and post published.
    static func notificationSuccess() {
        let g = UINotificationFeedbackGenerator()
        g.prepare()
        g.notificationOccurred(.success)
    }

    // MARK: Selection

    /// Selection tick — use for feed mode switch and new-posts pill appear.
    static func selectionChanged() {
        let g = UISelectionFeedbackGenerator()
        g.prepare()
        g.selectionChanged()
    }
}

// MARK: - Sound

/// Opt-in sound effects for Olas.
///
/// Gated by the `"soundEffectsEnabled"` UserDefaults key (default `false`) which mirrors
/// the in-app toggle in Settings → Appearance → Sound Effects.
///
/// Uses `AVAudioSession.Category.ambient` so effects:
/// - respect the hardware silent/ringer switch,
/// - do not interrupt background music or podcasts.
///
/// Audio assets ("ShutterSoft.caf", "ZapChime.caf") are **optional**. If an asset is
/// absent from the app bundle the call is a silent no-op. Real asset files can be added
/// later without changing any call-site code.
@MainActor
enum OlasSound {

    static var isEnabled: Bool {
        UserDefaults.standard.bool(forKey: "soundEffectsEnabled")
    }

    /// Soft camera shutter — plays on successful post publish.
    static func shutterSoft() {
        guard isEnabled else { return }
        play(named: "ShutterSoft", ext: "caf")
    }

    /// Warm chime — plays when a zap is sent.
    static func zapChime() {
        guard isEnabled else { return }
        play(named: "ZapChime", ext: "caf")
    }

    // MARK: Private

    // Retained across calls so the player is not deallocated mid-playback.
    private static var player: AVAudioPlayer?

    private static func play(named name: String, ext: String) {
        guard let url = Bundle.main.url(forResource: name, withExtension: ext) else {
            // Asset not yet bundled — silent no-op by design.
            return
        }
        do {
            try AVAudioSession.sharedInstance().setCategory(.ambient, mode: .default)
            let p = try AVAudioPlayer(contentsOf: url)
            p.prepareToPlay()
            p.play()
            player = p // retain until playback ends
        } catch {
            // Silently ignore playback failures (e.g., Simulator without audio support).
        }
    }
}
